### PR TITLE
delete: Remove both image stream spec and status tags

### PR DESCRIPTION
**Pull request by kargakis**
@smarterclayton @soltysh reopening this for discussion

### CODE REVIEW

File: pkg/image/registry/imagestream/etcd/etcd.go [hunk 1]
```diff
@@ -91,6 +91,17 @@ func (r *REST) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object, boo
 	return r.store.Update(ctx, obj)
 }
 
+// UpdateAll updates both the image stream specification and status.
+func (r *REST) UpdateAll(ctx kapi.Context, obj runtime.Object) (runtime.Object, bool, error) {
+	oldStrategy := r.store.UpdateStrategy
+	r.store.UpdateStrategy = imagestream.NewAllStrategy(oldStrategy.(imagestream.Strategy))
+	defer func() {
+		r.store.UpdateStrategy = oldStrategy
```
Comment (soltysh) [inline pkg/image/registry/imagestream/etcd/etcd.go]: https://github.com/openshift/origin/pull/3982#discussion_r42655808 is the reason for re-review.

**Comment by kargakis** [comment]
@smarterclayton ping

File: pkg/image/registry/imagestream/etcd/etcd.go [hunk 2]
```diff
@@ -91,6 +91,15 @@ func (r *REST) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object, boo
 	return r.store.Update(ctx, obj)
 }
 
+// UpdateAll updates both the image stream specification and status.
+func (r *REST) UpdateAll(ctx kapi.Context, obj runtime.Object) (runtime.Object, bool, error) {
+	oldStrategy := r.store.UpdateStrategy
+	r.store.UpdateStrategy = imagestream.NewAllStrategy(oldStrategy.(imagestream.Strategy))
```
Comment (smarterclayton) [inline pkg/image/registry/imagestream/etcd/etcd.go]: You can't do this - it's illegal to change anything in `r` during an edit.  You have to have a separate object to manage that.

Comment (liggitt) [inline pkg/image/registry/imagestream/etcd/etcd.go]: remember this object is shared between all requests made to the server

Comment (kargakis) [inline pkg/image/registry/imagestream/etcd/etcd.go]: ok, updated to use a separate store for updating both spec and status

File: pkg/cmd/server/origin/master.go [hunk 3]
```diff
@@ -372,8 +372,8 @@ func (c *MasterConfig) GetRestStorage() map[string]rest.Storage {
 
 	imageStorage := imageetcd.NewREST(c.EtcdHelper)
 	imageRegistry := image.NewRegistry(imageStorage)
-	imageStreamStorage, imageStreamStatusStorage := imagestreametcd.NewREST(c.EtcdHelper, imagestream.DefaultRegistryFunc(defaultRegistryFunc), subjectAccessReviewRegistry)
-	imageStreamRegistry := imagestream.NewRegistry(imageStreamStorage, imageStreamStatusStorage)
+	imageStreamStorage, imageStreamStatusStorage, isAllStorage := imagestreametcd.NewREST(c.EtcdHelper, imagestream.DefaultRegistryFunc(defaultRegistryFunc), subjectAccessReviewRegistry)
```
Comment (smarterclayton) [inline pkg/cmd/server/origin/master.go]: Awkward name, call it internalImageStreamStorage

File: pkg/image/registry/imagestream/etcd/etcd.go [hunk 4]
```diff
@@ -109,3 +112,17 @@ func (r *StatusREST) New() runtime.Object {
 func (r *StatusREST) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object, bool, error) {
 	return r.store.Update(ctx, obj)
 }
+
+// AllREST implements the REST endpoint for changing both the spec and status of an image stream.
+type AllREST struct {
```
Comment (smarterclayton) [inline pkg/image/registry/imagestream/etcd/etcd.go]: InternalREST

File: pkg/image/registry/imagestream/strategy.go [hunk 5]
```diff
@@ -421,3 +421,26 @@ type DefaultRegistryFunc func() (string, bool)
 func (fn DefaultRegistryFunc) DefaultRegistry() (string, bool) {
 	return fn()
 }
+
+// AllStrategy implements behavior for updating both the spec and status
+// of an image stream
+type AllStrategy struct {
```
Comment (smarterclayton) [inline pkg/image/registry/imagestream/strategy.go]: InternalStrategy

**Comment by smarterclayton** [comment]
Just the rename, otherwise looks fine.

File: pkg/cmd/server/origin/master.go (hunk 3 above)
Comment (kargakis) [inline pkg/cmd/server/origin/master.go]: ok

File: pkg/image/registry/imagestream/etcd/etcd.go (hunk 4 above)
Comment (kargakis) [inline pkg/image/registry/imagestream/etcd/etcd.go]: ok

File: pkg/image/registry/imagestream/strategy.go (hunk 5 above)
Comment (kargakis) [inline pkg/image/registry/imagestream/strategy.go]: ok

**Comment by kargakis** [comment]
@soltysh as well, fyi

**Comment by soltysh** [comment]
LGTM.

**Comment by kargakis** [comment]
[merge]

**Pull request merged**